proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,135 @@
+3.  a) Returning the top-k retrieved similar documents, if only one retrieval method was used.
+    b) Returning an aggregated list of similar documents retrieved by both retrievers, if two retrieval methods were chosen. The results are aggregated based on the reciprocal rank fusion score, though it is also possible to keep the original score (if one of the retrievers assigned the document a relevance score) in the metadata.
+    Reciprocal rank fusion is calculated as 1 / (k + index)
+    k is set to 61 in reciprocal rank fusion based on a University of Waterloo paper (co-authored with Google) called "Reciprocal Rank Fusion outperforms Condorcet and individual Rank Learning Methods" [https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf] where k=60 was suggested, and 1 was added as python lists are 0-based and the paper used 1-based ranking.
+4. Returning the top-k similar files, based on an aggregated relevance score of all documents pertaining to the file. In the results the user sees the most relevant document from each file.
+We can also provide the user a choice to have the values of the metadata key used for aggregation (e.g. file IDs) returned instead of the most relevant documents, e.g. via an additional optional param "output" set by the user to "top_document" or "file_aggregation_key", with "top_document" being default since it looks more similar to the output of all pipelines (as haystack pipelines usually return natural language responses).
+
+FileSimilarityRetriever has the following parameters:
+- document_store: The document store that the retriever should retrieve from.
+- file_aggregation_key: The meta data key that should be used to aggregate documents to the file level.
+- primary_retriever: First retriever.
+- secondary_retriever (optional): Second retriever (if applicable).
+- keep_original_score (optional): Set this to store the original relevance score of the returned document in the document's meta field (the documents' scores get replaced in the FileSimilarityRetriever output with the reciprocal rank fusion score).
+- top_k: How many documents to return.
+- max_query_len: How many characters can be in a query document. The documents exceeding this limit will be cut off. This was added because in some cases BM25Retriever threw an error if the query exceeded the `max_clause_count` search setting (https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-settings.html)
+- max_num_queries (optional): The maximum number of queries that should be run for a single file. If the number of query documents exceeds this limit, the query documents will be split into n parts so that n < max_num_queries and every nth document will be kept.
+- use_existing_embeddings: Whether to re-use the existing embeddings from the index. To optimize speed for the file similarity retrieval you should set this parameter to `True`. This way the FileSimilarityRetriever can run on the CPU.
```

Comment:
Is there any reason you would want to set this to `False`? You would probably also not recompute the document embeddings and only the query embeddings so it would also not allow changing embedding models

Reply:
I've ran it now with different retrieval options and with this param set to true & false, and indeed didn't find much reason to use "false" there, the param could be removed and the retriever could always re-use embeddings where possible I think. @mathislucka do you have an opinion on this?

Reply:
You might want to use a different embedding model or play with multiple ones quickly. The initial retrieval only uses the filter retriever so it's pretty independent which embedding model you use.